rp2040-hal/src/clocks/mod.rs

```diff
@@ -0,0 +1,213 @@
+}
+/// For clocks that can have ... itself (?) as a source (is that the "glitchless mux" ?)
+pub trait SelfAuxClockSource {
+    /// Set ...
+    fn set_self_aux_src(&mut self);
+}
+/// For clocks that can have the Reference Clock as source.
+pub trait ClockREFClockSource {
+    /// Set Reference Clock as
+    fn set_clkref_src(&mut self);
+}
+/// For clocks that can have the System Clock as an auxilliary source.
+pub trait ClockSYSClockAuxSource {
+    /// Set System Clock as source.
+    fn set_clksys_auxsrc(&mut self);
+}
+/// For clocks that can have XOSC as an auxilliary source.
+pub trait XOSCClockAuxSource {
+    /// Set XOSC as auxilliary source.
+    fn set_xosc_auxsrc(&mut self);
+}
+/// For clocks that can have ROSC as an auxilliary source.
+pub trait ROSCClockAuxSource {
+    /// Set ROSC as auxilliary source.
+    fn set_rosc_auxsrc(&mut self);
+}
+/// For clocks that can have ROSC_PH as an auxilliary source.
+pub trait ROSCPHClockAuxSource {
+    /// Set ROSC_PH as auxilliary source.
+    fn set_rosc_ph_auxsrc(&mut self);
+}
+/// For clocks that can have PLL_USB as an auxilliary source.
+pub trait PLLUSBClockAuxSource {
+    /// Set PLL_USB as auxilliary source.
+    fn set_pll_usb_auxsrc(&mut self);
+}
+/// For clocks that can have PLL_SYS as an auxilliary source.
+pub trait PLLSYSClockAuxSource {
+    /// Set PLL_SYS as auxilliary source.
+    fn set_pll_sys_auxsrc(&mut self);
+}
+/// For clocks that can have gpin0 as an auxilliary source.
+pub trait Gpin0ClockAuxSource {
+    /// Set clock to be received from gpin0 (auxilliary)
+    fn set_gpin0_auxsrc(&mut self);
+}
+/// For clocks that can have gpin1 as an auxilliary source.
+pub trait Gpin1ClockAuxSource {
+    /// Set clock to be received from gpin1
+    fn set_gpin1_auxsrc(&mut self);
+}
+
+/// For clocks having a generator.
+pub trait ClockGenerator {
+    /// Enables the clock.
+    fn enable(&mut self);
+
+    /// Disables the clock.
+    fn disable(&mut self);
+
+    /// Kills the clock.
+    fn kill(&mut self);
+}
+
+/// Reference Clock
+pub struct ReferenceClock {
+    shared_dev: ShareableClocks,
+}
+xosc_source!(ReferenceClock, clk_ref_ctrl);
+rosc_source!(ReferenceClock, clk_ref_ctrl);
+selfaux_source!(ReferenceClock, clk_ref_ctrl, clksrc_clk_ref_aux);
+gpin0_auxsource!(ReferenceClock, clk_ref_ctrl);
+gpin1_auxsource!(ReferenceClock, clk_ref_ctrl);
+pll_usb_auxsource!(ReferenceClock, clk_ref_ctrl);
+int_division!(ReferenceClock, clk_ref_div, u8);
+
+/// System Clock
+pub struct SystemClock {
+    shared_dev: ShareableClocks,
+}
+impl SystemClock {
+    /// WIP - Helper function to reset source (blocking)
+    pub fn reset_source_await(&self) {
+        let shared_dev = unsafe { self.shared_dev.get() };
+
+        shared_dev.clk_sys_ctrl.write(|w| {
+            w.src().clear_bit();
+            w
+        });
+
+        self.await_select(0x1);
+    }
+
+    /// WIP - Helper function to select new source (blocking)
+    pub fn await_select(&self, clock: u8) {
```

Comment:
This is currently a bit of a pain to use from a user perspective. I don't need to care what number corresponds to what clock source when setting it, but I do need to know that to await it being selected.

You may have a better idea on how to improve this, but one potential way is to have the clock source switches that can be awaited just return a marker type that you can call await on (or just throw it away if you don't care)

Reply:
Agreed. I like the idea of the Marker type !